src/lib/transition_frontier/persistent_frontier/database.ml

```diff
@@ -0,0 +1,330 @@
+      Batch.set batch ~key:Db_version ~data:version ;
+      Batch.set batch ~key:(Transition root_state_hash) ~data:root_transition ;
+      Batch.set batch ~key:(Arcs root_state_hash) ~data:[] ;
+      Batch.set batch ~key:Root ~data:(Root_data.Minimal.of_limited root_data) ;
+      Batch.set batch ~key:Best_tip ~data:root_state_hash ;
+      Batch.set batch ~key:Frontier_hash ~data:base_hash )
+
+let add t ~transition =
+  let parent_hash = External_transition.Validated.parent_hash transition in
+  let {With_hash.hash; data= raw_transition}, _ =
+    External_transition.Validated.erase transition
+  in
+  let%map () =
+    Result.ok_if_true
+      (mem t.db ~key:(Transition parent_hash))
+      ~error:(`Not_found `Parent_transition)
+  in
+  let parent_arcs = get_if_exists t.db ~key:(Arcs parent_hash) ~default:[] in
+  Batch.with_batch t.db ~f:(fun batch ->
+      Batch.set batch ~key:(Transition hash) ~data:raw_transition ;
+      Batch.set batch ~key:(Arcs parent_hash) ~data:(hash :: parent_arcs) )
+
+let move_root t ~new_root ~garbage =
+  let open Root_data.Minimal.Stable.V1 in
+  let%bind () =
+    Result.ok_if_true
+      (mem t.db ~key:(Transition new_root.hash))
+      ~error:(`Not_found `New_root_transition)
+  in
+  let%map old_root =
+    get t.db ~key:Root ~error:(`Not_found `Old_root_transition)
+  in
+  (* TODO: Result compatible rocksdb batch transaction *)
+  Batch.with_batch t.db ~f:(fun batch ->
+      Batch.set batch ~key:Root ~data:new_root ;
+      List.iter (old_root.hash :: garbage) ~f:(fun node_hash ->
+          (* because we are removing entire forks of the tree, there is
+       * no need to have extra logic to any remove arcs to the node
+       * we are deleting since there we are deleting all of a node's
+       * parents as well
+       *)
+          Batch.remove batch ~key:(Transition node_hash) ;
+          Batch.remove batch ~key:(Arcs node_hash) ) ) ;
+  old_root.hash
+
+let get_transition t hash =
+  let%map transition =
+    get t.db ~key:(Transition hash) ~error:(`Not_found (`Transition hash))
+  in
+  (* this transition was read from the database, so it must have been validated already *)
+  let (`I_swear_this_is_safe_see_my_comment validated_transition) =
+    External_transition.Validated.create_unsafe transition
+  in
+  validated_transition
+
+let get_arcs t hash =
+  get t.db ~key:(Arcs hash) ~error:(`Not_found (`Arcs hash))
+
+let get_root t = get t.db ~key:Root ~error:(`Not_found `Root)
+
+let get_root_hash t =
+  let%map root = get_root t in
+  root.hash
+
+let get_best_tip t = get t.db ~key:Best_tip ~error:(`Not_found `Best_tip)
+
+let set_best_tip t hash =
+  let%map old_best_tip_hash = get_best_tip t in
+  (* no need to batch because we only do one operation *)
+  set t.db ~key:Best_tip ~data:hash ;
+  old_best_tip_hash
+
+let get_frontier_hash t =
+  get t.db ~key:Frontier_hash ~error:(`Not_found `Frontier_hash)
+
+(* TODO: bundle together with other writes using batch? *)
```

Comment:
Yes, we should ensure that all operations that we receive involving a diff should be atomic.

Reply:
#3738